engines/router/missionctl/log/resultlog/resultlog.go

```diff
@@ -42,131 +54,141 @@ var ResultLogKeys = struct {
+		} else {
+			// Process the response body
+			uncompressedData, err := uncompressHTTPBody(resp.header, resp.body)
+			if err != nil {
+				logger.Errorf("Error occurred when reading %s response body: %s",
+					resp.key, err.Error())
+				AddResponse(logEntry, resp.key, "", nil, err.Error())
+			} else {
+				logger.Debugw("Logging response", "respBody", string(uncompressedData))
+				// Format the response header
+				responseHeader := FormatHeader(resp.header)
+				AddResponse(logEntry, resp.key, string(uncompressedData), responseHeader, "")
+			}
+		}
+	}
+
+	logger.Debugw("Received all response from mcRespCh")
+	// Log the responses. If an error occurs in logging the result to the
+	// configured result log destination, log the error.
+	if err = rl.logEntry(logEntry); err != nil {
+		logger.Errorf("Result Logging Error: %s", err.Error())
 	}
-	return kvPairs, nil
 }
 
-// AddResponse adds the per-component response/error info to the TuringResultLogEntry
-func (logEntry *TuringResultLogEntry) AddResponse(key string, body string, header map[string]string, err string) {
-	responseRecord := &turing.Response{
-		Header:   header,
-		Response: body,
-		Error:    err,
+// LogTuringRouterRequestError logs the given turing request id and the error data
+func (rl *ResultLogger) LogTuringRouterRequestError(ctx context.Context, err *errors.TuringError) {
+	logger := log.WithContext(ctx)
+	defer func() {
+		_ = logger.Sync()
+	}()
+	logger.Errorw("Turing Request Error",
+		"error", err.Message,
+		"status", err.Code,
+	)
+}
+
+// SendResponseToLogChannel copies the response from the turing router to the given channel
+// as a RouterResponse object
+func (rl *ResultLogger) SendResponseToLogChannel(
+	ctx context.Context,
+	ch chan<- RouterResponse,
+	key string,
+	r mchttp.Response,
+	httpErr *errors.TuringError,
+) {
+	var data []byte
+
+	// if http error is not nil, use error as response
+	if httpErr != nil {
+		ch <- RouterResponse{
+			key: key,
+			err: httpErr.Message,
+		}
+		return
 	}
-	switch key {
-	case ResultLogKeys.Experiment:
-		logEntry.resultLogMessage.Experiment = responseRecord
-	case ResultLogKeys.Enricher:
-		logEntry.resultLogMessage.Enricher = responseRecord
-	case ResultLogKeys.Router:
-		logEntry.resultLogMessage.Router = responseRecord
-	case ResultLogKeys.Ensembler:
-		logEntry.resultLogMessage.Ensembler = responseRecord
+
+	data = r.Body()
+	if data == nil {
+		// Error in logging method, doesn't have to be propagated. Simply log the error.
+		logger := log.WithContext(ctx)
+		defer func() {
+			_ = logger.Sync()
+		}()
+		logger.Errorf("Error occurred when reading data from %s", key)
+	}
+	// Copy to channel
+	ch <- RouterResponse{
+		key:    key,
+		header: r.Header(),
+		body:   data,
 	}
 }
 
-// NewTuringResultLogEntry returns a new TuringResultLogEntry object with the given context
+func (rl *ResultLogger) logEntry(log *turing.TuringResultLogMessage) error {
+	log.RouterVersion = rl.appName
```

Comment:
routerversion is being appended at this layer as the resultlogger owns the logic on how it should be populated and not the server.